File: 0133-clone-graph/0133-clone-graph.js
/**
 * // Definition for a _Node.
 * function _Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {_Node} node
 * @return {_Node}
 */
var cloneGraph = function (root) {
    if (!root) return null;

    const stack = [root];
    const visited = new Map();
    const clonedRootNode = new Node(root.val);
    visited.set(root, clonedRootNode);

    while (stack.length > 0) {
        const curr = stack.pop();
        const clonedNode = visited.get(curr);

        for (const n of curr.neighbors) {
            if (!visited.has(n)) {
                stack.push(n);
                visited.set(n, new Node(n.val));
            }
            clonedNode.neighbors.push(visited.get(n));
        }
    }

    return clonedRootNode;
};

File: 0133-clone-graph/0133-clone-graph.py
from collections import deque
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, root: Optional['Node']) -> Optional['Node']:
        if root == None:
            return None

        q = deque([root])
        clonedRootNode = Node(root.val)
        visited = {}
        visited[root] = clonedRootNode

        while len(q) > 0:
            curr = q.popleft()
            clonedCurr = visited[curr]

            for n in curr.neighbors:
                if n not in visited:
                    q.append(n)
                    visited[n] = Node(n.val)
                clonedCurr.neighbors.append(visited[n])
        
        return clonedRootNode

